man/man3/pnet_setup.3
.TH pnet_setup 3  "version 1.0.3" "Pnet programmers manual"
.SH NAME
pnet_setup \- Initialize communication pattern
.SH SYNOPSIS
.nf
\fB#include <mpi.h>
#include "pnet.h"

int pnet_setup(Pnet\fP *\fIpnet\fB);\fP
.fi
.SH DESCRIPTION
Function initializes communication pattern necessary for the 
distributed network. This is the core function of the Pnet
library. Pnet has to be initialized by \fBpnet_init\fP and connections have to be added by calls of \fBpnet_addconnect\fP before \fBpnet_setup\fP can be called. Data have to be exchanged by invoking the \fBpnet_exchg\fP
macro. After calling this function connections must not be added.
The argument of \fBpnet_setup\fP is:
.TP
.I pnet
Pointer to initialized Pnet struct 
.SH RETURN VALUE
Upon successful completion \fBPNET_OK\fP is returned.
.SH ERRORS
.TP
.B PNET_ALLOC_ERR
Cannot allocate memory.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
pnet_exchg(3), pnet_init(3), pnet_addconnect(3), pnet_free(3),  pnet_lo(3), pnet_hi(3), pnet_nodes(3)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

## [5.7.6] - 2023-09-01

### Added

- reallocation of N added in `allometry_tree.c` if tree height exceeds the maximum height.
- reproduction costs for nitrogen added to `turnover_tree.c`.

### Changed

- check for `sum>1` for warnings has been replaced by `sum>1+epsilon` in `getlanduse()`.

### Fixed

- missing `#include "grass.h"` added in `sowingcft.c`.
- arguments of `setaside` calls in `annual_woodplantation.c` and `annual_biomass_tree.c` have been corrected.
- update of `LITFALLN` corrected in `timber_burn.c`.
- output written to correct `N2O_DENIT_*` file in `denitrification.c`.
- missing `reservoirfrac` added to scaling for outputs in `daily_agriculture_grass.c` and `daily_agriculture_tree.c`.


## [5.7.5] - 2023-08-23

### Added

- outputs:
  - IRRIG_STOR
  - RIVERVOL
  - SWC_VOL
 
### Changed:

- unit of output RES_STORAGE changed from hm3 to dm3 to be in line with IRRIG_STOR, RIVERVOL, LAKEVOL, SWC_VOL

### Fixed

- check for land-use fractions to not exceed 100% after scaling with `landfrac`, re-scale to 100%, print warning see [!173](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/merge_requests/173)
- add turnover before allocation in cultcftstand to prevent sporadic C balance errors when running with intercrops, see[!175](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/merge_requests/175)
- number of wet days not read if daily precipitation input is used. A warning is additionally printed to set `"random_prec"` to false.
- boolean `"river_routing"` is read before `"with_lakes"` flag to avoid uninitialized variable.


## [5.7.4] - 2023-08-17

### Changed

- changed quadratic soil evaporation function to sigmoid form and a minimum amount of evaporation of 5% of the available energy, following [Sun et al. 2013](http://dx.doi.org/10.1080/17538947.2013.783635), as described in [!168](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/merge_requests/168)

## [5.7.3] - 2023-08-16

### Fixed

- correct accounting of blue water transpiration on agriculture stands in outputs
- missing variable initialization for bioenergy grass
- landuse change from irrigated to rainfed setaside moved area to wrong target setaside


## [5.7.2] - 2023-08-11

### Changed:

- removed timestep attributes (monthly/annual) in all output descriptions that are not timestep specific

### Added 

- outputs:
  - PFT_WATER_DEMAND: PFT specific water demand
  - RD: dark respiration
  - NDEPOS: total N deposition as sum of NH4 and NO3 deposition

## [5.7.1] - 2023-08-10

### Fixed

- wrong use of `&f` instead of `f`, in `fscanparam.c`. Bug was introduced in version 5.6.26 and caused that parameter files could not be read properly

## [5.7.0] - 2023-08-10

### Added

- land fraction can now be read from file, with a new flag `"landfrac_from_file"` and a new input `"landfrac"`
- If setting `"landfrac_from_file : true"` cell areas in LPJmL are scaled down by land fraction from input file. If land fraction is zero, land fraction is set to `minlandfrac` and a warning is printed.
- new flag `"with_lakes"` defined
- new outputs `"terr_area"`, `"land_area"` and `"lake_area"` defined writing the terrestrial area (land and lakes),  land area and lake area of each cell
- new parameters `"minlandfrac"` and `"tinyfrac"` added to lpjparams.js
- option `-area filename` added to the `printglobal` utility to scale with the cell areas read from the file specified.

### Changed

- Lake fraction and land-use fractions are rescaled by the reciprocal of land fraction. If lake fraction is greater than one, lake fraction is set to one and a warning is printed.
- Lakes can be enabled without river routing by setting `"with_lakes" : true`. Variable `dmass_lake` is always written to restart file.
- lake fraction is read in utility `lpjprint` and printed
- flag `"shuffle_spinup_climate"` only read if `"nspinup"` is non-zero.
- Terrestrial area and harvest fraction are read in utility `printharvest`. Output format changed to CSV.

## [5.6.27] - 2023-08-08

### Removed

- removed large chunks of code in `allocation_daily_crop.c` that was inactive

### Fixed

- fixed output `PFT_VEGC` to not include `bm_inc.nitrogen` for crops in `daily_agriculture.c`
- fixed global flux summation for `flux.soil.carbon` and `flux.soil.nitrogen` (`+=` instead of `=`), bug was introduced in version 5.6.25

## [5.6.26] - 2023-08-07

### changed

- avoid daily leaf turnover for biomass grass and harvest leaf turnover as dead biomass at harvest event instead (reduces litter and increases harvest)
- introduce harvest dates depending on green/brown harvest instead of C threshold for biomass grass
- literature-based C/N ratios and fn_turnover for biomass grass depending on green/brown harvest configuration
- results and literature references can be found in issue [#275](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/275).

### Added

- introduced `#define` flag to change from green harvest (default) to brown harvest: `#define BMGR_BROWN`
- two options include green harvest (fresh biomass, high N content - usually used for biofuels) and brown harvest (dry biomass, high lignin content, low N content - usually used for combustion)

src/pnet/pnet_setup.c
/**************************************************************************************/
/**                                                                                \n**/
/**                     p  n  e  t  _  s  e  t  u  p  .  c                         \n**/
/**                                                                                \n**/
/**     MPI-parallelization of networks                                            \n**/
/**                                                                                \n**/
/**     Function initializes communication pattern necessary for the               \n**/
/**     distributed network. This is the core function for the Pnet                \n**/
/**     library. Data have to be exchanged by incoking the pnet_exchg              \n**/
/**     macro.                                                                     \n**/
/**                                                                                \n**/
/**     Implementation is described in:                                            \n**/
/**     W. von Bloh, 2008. Sequential and Parallel Implementation of               \n**/
/**     Networks. In P. beim Graben, C. Zhou, M. Thiel, and J. Kurths              \n**/
/**     (eds.), Lectures in Supercomputational Neuroscience, Dynamics in           \n**/
/**     Complex Brain Networks, Springer, 279-318.                                 \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#ifdef USE_MPI
#include <mpi.h>
#endif
#include "types.h"
#include "pnet.h"

static int compare(const int *a,const int *b)
{
  /* compare function to sort integers in ascending order, used by qsort() */
  return *a-*b;
} /* of 'compare' */

int pnet_setup(Pnet *pnet /**< Pointer to Pnet structure */
              )           /** \return error code        */
{
  int *lo,*hi;
  int i,j,k,*index,*in,size,slice,rem,task,insize;
#ifdef USE_MPI
  MPI_Aint extent;
#endif
  if(pnet==NULL)
    return PNET_NULL_PTR_ERR;
  slice=pnet->n/pnet->ntask;
  rem=pnet->n % pnet->ntask;
  lo=newvec(int,pnet->ntask);
  if(lo==NULL)
    return PNET_ALLOC_ERR;
  hi=newvec(int,pnet->ntask);
  if(hi==NULL)
  {
    free(lo);
    return PNET_ALLOC_ERR;
  }
  /* calculate lower and upper bounds for all tasks */
  for(i=0;i<pnet->ntask;i++)
  {
    lo[i]=i*slice;
    hi[i]=(i+1)*slice-1;
    if(i<rem)
    {
       lo[i]+=i;
       hi[i]+=i+1;
    }
    else
    {
      lo[i]+=rem;
      hi[i]+=rem;
    }
  }
  for(i=0;i<pnet->ntask;i++)
    pnet->outlen[i]=pnet->inlen[i]=0;
  /* determine total length of connection lists */
  size=0;
  pnet_foreach(pnet,i)
    size+=pnet->connect[i].n;
  in=newvec(int,size);
  if(in==NULL)
  {
    /* error allocating memory, exit with error code */
    free(lo);
    free(hi);
    pnet->outindex=NULL;
    pnet->inbuffer=pnet->outbuffer=NULL;
    return PNET_ALLOC_ERR;
  }
  if(size) /* are connection lists not all empty? */
  {
    /* concatenate connection lists to array in */
    k=0;
    for(i=pnet->lo;i<=pnet->hi;i++)
      for(j=0;j<pnet->connect[i].n;j++)
        in[k++]=pnet->connect[i].index[j];
    /* sort in ascending order */
    qsort(in,size,sizeof(int),(int (*)(const void *,const void *))compare);
    /* delete duplicates */
    insize=1;
    for(i=1;i<size;i++)
      if(in[i]!=in[i-1])
      {
        /* let in[i] in array if distinct from in[i-1] */
        in[insize]=in[i];
        insize++;
      }
    task=0;
    /* calculate inlen vector */
    pnet->inlen[task]=0;
    i=0;
    while(i<insize)
      if(in[i]<=hi[task])
      {
        /* in[i] is within lower and upper bound of task */
        pnet->inlen[task]++;
        i++;
      }
      else
      {
        /* move to next task */
        task++;
        pnet->inlen[task]=0;
      }
    /* create index vectors for array in */
    for(i=pnet->lo;i<=pnet->hi;i++)
      for(j=0;j<pnet->connect[i].n;j++)
      {
        /* performing binary search for index in in vector */
        index=(int *)bsearch(pnet->connect[i].index+j,in,insize,sizeof(int),
                             (int (*)(const void *,const void *))compare);
        /* replace entry in connection list by index */
        pnet->connect[i].index[j]=index-in;
      }
  }
  else
    insize=0;
#ifdef USE_MPI
  /* distribute information of inlen vectors to all tasks */
  MPI_Alltoall(pnet->inlen,1,MPI_INT,pnet->outlen,1,MPI_INT,pnet->comm);
#else
  /* simple copy for sequential code */
  pnet->outlen[0]=pnet->inlen[0];
#endif
  /* calculate displacement vectors */
  pnet->indisp[0]=pnet->outdisp[0]=0;
  for(k=1;k<pnet->ntask;k++)
  {
    pnet->indisp[k]=pnet->inlen[k-1]+pnet->indisp[k-1];
    pnet->outdisp[k]=pnet->outlen[k-1]+pnet->outdisp[k-1];
  }
#ifdef DEBUG
  printf("in %d:",pnet->taskid);
  k=0;
  for(i=0;i<insize;i++)
  {
    if(pnet->indisp[k]+pnet->inlen[k]==i)
    {
      printf("|");
      k++;
    }
    printf(" %d",in[i]);
  }
  printf("\n");
#endif
  /* calculate length of output buffer */
  pnet->outsize=0;
  for(i=0;i<pnet->ntask;i++)
    pnet->outsize+=pnet->outlen[i];
  /* allocate output index vector */
  pnet->outindex=newvec(int,pnet->outsize);
  if(pnet->outindex==NULL)
  {
    free(lo);
    free(hi);
    free(in);
    pnet->inbuffer=pnet->outbuffer=NULL;
    return PNET_ALLOC_ERR;
  }
#ifdef USE_MPI
  /* distribute index vectors */
  MPI_Alltoallv(in,pnet->inlen,pnet->indisp,MPI_INT,pnet->outindex,pnet->outlen,
                pnet->outdisp,MPI_INT,pnet->comm);
#else
  for(i=0;i<pnet->inlen[0];i++)
    pnet->outindex[i]=in[i];
#endif
#ifdef DEBUG
  printf("out %d:",pnet->taskid);
  k=0;
  for(i=0;i<pnet->outsize;i++)
  {
    if(pnet->outlen[k]+pnet->outdisp[k]==i)
    {
      printf("|");
      k++;
    }
    printf(" %d",pnet->outindex[i]);
  }
  printf("\n");
#endif
  /* Allocation of input and output buffer */
#ifdef USE_MPI
  MPI_Type_extent(pnet->type,&extent); /* calculate size of datatype */
  pnet->outbuffer=malloc(extent*pnet->outsize);
  pnet->inbuffer=malloc(extent*insize);
#else
  pnet->outbuffer=malloc(pnet->size*pnet->outsize);
  pnet->inbuffer=malloc(pnet->size*insize);
#endif
  free(in);
  free(lo);
  free(hi);
  return (pnet->outbuffer==NULL || pnet->inbuffer==NULL) ? PNET_ALLOC_ERR : PNET_OK;
} /* of 'pnet_setup' */
